Pyrado/pyrado/algorithms/meta/pddr.py

```diff
@@ -0,0 +1,363 @@
+        """ Update the policy's (and value functions') parameters based on the collected rollout data. """
+        obss = []
+        losses = []
+        for t in range(self.num_teachers):
+            concat_ros = StepSequence.concat(kwargs["rollouts"][t])
+            concat_ros.torch(data_type=to.get_default_dtype())
+            obss.append(concat_ros.get_data_values("observations")[: self.min_steps])
+
+        # Train student
+        for epoch in range(self.num_epochs):
+            self.optimizer.zero_grad()
+
+            loss = 0
+            for t_idx, teacher in enumerate(self.teachers):
+                s_dist = self.expl_strat.action_dist_at(self.policy(obss[t_idx]))
+                s_act = s_dist.sample()
+                t_dist = self.teacher_expl_strat[t_idx].action_dist_at(teacher(obss[t_idx]))
+
+                l = self.teacher_weights[t_idx] * self.criterion(t_dist.log_prob(s_act), s_dist.log_prob(s_act))
+                loss += l
+                losses.append([t_idx, l.item()])
+            print(f"Epoch {epoch} Loss: {loss.item()}")
+            loss.backward()
+            self.optimizer.step()
+
+    def save_snapshot(self, meta_info: dict = None):
+        super().save_snapshot(meta_info)
+
+        pyrado.save(self.policy, "policy.pt", self.save_dir)
+
+        if meta_info is None:
+            # This algorithm instance is not a subroutine of another algorithm
+            pyrado.save(self.env_real, "env.pkl", self.save_dir)
+
+    def load_teachers(self):
+        # Get the experiment's directory to load from
+        ex_dir = ask_for_experiment()
+        self.load_teacher_experiment(ex_dir)
+        if len(self.teachers) < self.num_teachers:
+            print(
+                f"You have loaded {len(self.teachers)} teachers - load at least {self.num_teachers - len(self.teachers)} more!"
+            )
+            self.load_teachers()
+
+    def load_teacher_experiment(self, exp: Experiment):
+        """ Load teachers from PDDRTeachers experiment. """
+        _, _, extra = load_experiment(exp)
+        self.unpack_teachers(extra)
+
+    def unpack_teachers(self, extra):
+        """ Unpack teachers from PDDRTeachers experiment. """
+        self.teachers.extend(extra["policies"])
+        self.teacher_envs.extend(extra["envs"])
+        self.teacher_expl_strat.extend(extra["expl_strats"])
+        self.teacher_critics.extend(extra["critics"])
+        self.ex_dirs.extend(extra["ex_dirs"])
+
+    def prune_teachers(self):
+        """ Prune teachers to only use the first num_teachers of them. """
+        self.teachers = self.teachers[: self.num_teachers]
+        self.teacher_envs = self.teacher_envs[: self.num_teachers]
+        self.teacher_expl_strat = self.teacher_expl_strat[: self.num_teachers]
+        self.teacher_critics = self.teacher_critics[: self.num_teachers]
+        self.ex_dirs = self.ex_dirs[: self.num_teachers]
+
+
+class PDDRTeachers(Algorithm):
```

Comment:
I am not a fan on the additional class for PDDR teacher training. Why? Because you create an algorithm which is actually not really a disctinct algorithm. What you want is to run a policy optimitation subroutine on one exact domain parameter set (I mean an env that has been set to it). This is precisely what SPOTA does when computing the references. I would much prefer if you put this into a function like it is done there. I know that the loading logic would need to me moved to PDDR then.
Is there a strong argument for keeping it like this?